### ISSUE TITLE
The network does'nt train: every input produce the same output

### ISSUE THREAD

**Issue by Benedetta89** [BVLC/caffe#4300]
I have trained my net with a set of images 27x27 obtained from original ones of dimension 384x365. Each image has an associated label with a value that can be 0 or 1: I have a binary classification problem. I converted these images into an lmdb file for training. The net I used is the following one:

layer {
  name: "data"
  type: "Input"
  top: "data"
  input_param { shape: { dim: 1 dim: 3 dim: 27 dim: 27 } }
    include {
    phase: TEST
  } 
}

layer {
  name: "data"
  type: "Data"
  top: "data"
  transform_param {
crop_size: 27
  }
  data_param {
    
    source: "/home/deepcuda/Scrivania/caffe_ordinato/database27/image_train_lmdb/"
    batch_size: 256
    backend: LMDB
  }
      include {
    phase: TRAIN
  }   
  
}
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "data"
  top: "conv1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 64
    kernel_size: 3
    stride: 1
    pad:1
    weight_filler {
      type: "xavier"     
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "relu1"
  type: "ReLU"
  bottom: "conv1"
  top: "conv1"
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "conv1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv2"
  type: "Convolution"
  bottom: "pool1"
  top: "conv2"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 128
    kernel_size: 3
    stride: 1
    pad:1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "relu2"
  type: "ReLU"
  bottom: "conv2"
  top: "conv2"
}
layer {
  name: "pool2"
  type: "Pooling"
  bottom: "conv2"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv3"
  type: "Convolution"
  bottom: "pool2"
  top: "conv3"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 128
    kernel_size: 3
    stride: 1
    pad:1
    weight_filler {
      type: "xavier"
      
    }
    bias_filler {
      type: "constant"
      
    }
  }
}
layer {
  name: "relu3"
  type: "ReLU"
  bottom: "conv3"
  top: "conv3"
}
layer {
  name: "pool3"
  type: "Pooling"
  bottom: "conv3"
  top: "pool3"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}

layer {
  name: "fc4"
  type: "InnerProduct"
  bottom: "pool3"
  top: "fc4"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 512
   weight_filler {
      type: "gaussian" # initialize the filters from a Gaussian
      std: 0.01        # distribution with stdev 0.01 (default mean: 0)
    }

    bias_filler {
      type: "constant"
    }
  }
}

layer {
  name: "fc5"
  type: "InnerProduct"
  bottom: "fc4"
  top: "fc5"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 512
   weight_filler {
      type: "gaussian" # initialize the filters from a Gaussian
      std: 0.01        # distribution with stdev 0.01 (default mean: 0)
    }

    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "fc6"
  type: "InnerProduct"
  bottom: "fc5"
  top: "fc6"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 512
    weight_filler {
      type: "gaussian" # initialize the filters from a Gaussian
      std: 0.01        # distribution with stdev 0.01 (default mean: 0)
    }

    bias_filler {
      type: "constant"
    }
  }
}

layer {
  name: "fc7"
  type: "InnerProduct"
  bottom: "fc6"
  top: "fc7"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  param {
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 2
    weight_filler {
      type: "gaussian" # initialize the filters from a Gaussian
      std: 0.01        # distribution with stdev 0.01 (default mean: 0)
    }

    bias_filler {
      type: "constant"
    }
  }
}

layer {
  name: "loss"
  type: "Softmax"
  bottom: "fc7"
  top: "loss"
}

And the solver is this one:

# The train/test net protocol buffer definition

net: "/home/deepcuda/Scrivania/caffe_ordinato/model/deploy_articolo.prototxt" 

base_lr: 0.001     # begin training at a learning rate of 0.01 = 1e-2

lr_policy: "step" # learning rate policy: drop the learning rate in "steps"
                  # by a factor of gamma every stepsize iterations

gamma: 0.1        # drop the learning rate by a factor of 10
                  # (i.e., multiply it by a factor of gamma = 0.1)

stepsize: 10000  # drop the learning rate every 100K iterations

max_iter: 30000  # train for 350K iterations total

momentum: 0.9

#iter_size: 2   #ogni quanti batch aggiorna il gradiente

# snapshot intermediate results

snapshot: 5000

snapshot_prefix: "/home/deepcuda/Scrivania/caffe_ordinato/model/snapshot/" 

# solver mode: CPU or GPU

solver_mode: GPU

device_id:1

I trained my net with the command ./caffe train 
Then for the test phase I created a python script in wich i put into the network each single image 27x27  (The input to the TEST network is of type: "Input" { shape: { dim: 1 dim: 3 dim: 27 dim: 27 } })
and i carried out the prediction of the trained network for each of these images.
the script is like this:

import scipy
import numpy
from scipy import misc
import caffe
import sys
caffe_root = '/home/deepcuda/Scrivania/caffe_ordinato/'  
sys.path.insert(0, caffe_root + 'python')

####################################caricamento pathname################################
pathdir1="/home/deepcuda/Scrivania/dataset2/"
pathdir2="/home/deepcuda/Scrivania/dataset2/"
pathdir3="/home/deepcuda/Scrivania/dataset2/"
test_imgName=[]
test_maskName=[]
test_grtName=[]

for num in range(1,3): 
    string1=str(num)
    string2="_image.jpg"
    string=pathdir1+string1+string2     
    test_imgName.append(string)
    test_maskName.append(string)
    test_grtName.append(string)
    print(test_imgName[num-1])

test_original_images=numpy.zeros((2,3,584,565))
test_grt_images=numpy.zeros((2,3,584,565))
test_mask_images=numpy.zeros((2,3,584,565))

for num in range(1,3):
    test_original_images[num-1]=misc.imread(test_imgName[num-1])
    test_grt_images[num-1]=misc.imread(test_maskName[num-1])
    test_mask_images[num-1]=misc.imread(test_grtName[num-1])
    
print('test_original_images.shape')    
print(test_original_images.shape)    

caffe.set_mode_gpu()

model_def = caffe_root + 'model/deploy_articolo.prototxt'
model_weights = caffe_root + 'model/snapshot/_iter_30000.caffemodel'

net = caffe.Net(model_def,      # defines the structure of the model
                model_weights,  # contains the trained weights
                caffe.TEST)     # use test mode (e.g., don't perform dropout)
                

                                       
b=test_original_images[0:1,0:3,0:27,0:27]
print('b.shape')
print(b.shape)  
              

# copy the image data into the memory allocated for the net
net.blobs['data'].data[...] = b

#for i in range (0,10):

    ### perform classification
output = net.forward()

output_prob = output['loss']
print(output_prob)
print(output_prob.shape)

count=0
count_total=0
for i in range (0,319):
    for j in range (0,300):
        count_total=count_total+1
        b=test_original_images[0:1,0:1,i:i+27,j:j+27]    ##secondo indice da 0:3 a 0:1
        net.blobs['data'].data[...] = b
        output = net.forward()

        output_prob = output['loss']
        print(output_prob)
        print(output_prob.shape)
        if output_prob[0][0] > 0.49:
            count=count+1

THE PROBLEM IS THAT THE PROBABILITES ARE THE SAME FOR EACH INPUT IMAGE:
THE OUTPUT IS 
(1, 3, 27, 27)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
(1, 2)
[[ 0.4306114   0.56938857]]
....an so on (for all the images)

 How the problem caused?
Is it a training problem or it depends on how I took the predictions in the python script??
Please help me to understand

**Comment by seanbell** [comment]
It looks like you're not preprocessing the images, converting them from RGB to BGR order, and subtracting the mean?

Please ask for help with usage and debugging on the mailing list.

From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:
> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.

**Comment by Benedetta89** [comment]
Why Have I to convert them into BGR order?

**Comment by seanbell** [comment]
Well, if you trained using RGB order, then no you don't have to switch.  But if you used the standard caffe data layer, then it uses opencv which uses BGR order.  Python PIL and numpy/scipy generally use RGB order, so you need to swap.

It looks like you didn't subtract the mean at training time (which is a bad idea and makes learning less likely to converge), but if so you don't need to subtract the mean at test time.

Look at the preprocessing section (section 2) in http://nbviewer.jupyter.org/github/BVLC/caffe/blob/master/examples/00-classification.ipynb for a list of common preprocessing items done for ImageNet networks.

Please follow up on the mailing list; the issues page is not the right place to discuss help/usage.
